.github/workflows/test.yml

```diff
@@ -25,21 +25,64 @@ jobs:
         sudo adduser "$USER" lxd
         sudo -g lxd lxd init --auto
 
-    - name: Build charm
+    - name: Build charms
       run: |
         set -eux
         sudo -g lxd charmcraft build
 
+        # ancillary charm for testing purposes
+        sudo -g lxd charmcraft build --from examples/https-client
+
     - name: Bootstrap Juju
       run: |
         set -eux
         sudo -g lxd lxc network set lxdbr0 ipv6.address=none
         sudo -g lxd juju bootstrap lxd local
 
-    - name: Deploy the charm
+    - name: Exercice lxd-https relation
+      run: |
+        set -eux
+
+        # capture juju's logs on error
+        debug() {
+            rc="$?"
+            set +e
+            juju debug-log --replay --level DEBUG > juju-debug.log
+            exit "$rc"
+        }
+        trap debug ERR EXIT
+
+        juju deploy ./lxd_ubuntu-20.04-amd64.charm --num-units 1 --config sysctl-tuning=false --config kernel-hardening=false --config lxd-listen-https=true
+        juju deploy ./https-client_ubuntu-20.04-amd64.charm
+
+        # waiting is required before relating https-client with lxd
+        # as otherwise the relation establishment is done out of order
+        juju wait
```

Comment:
Can you check with @morphis or someone else who uses Juju more than we do on whether this is okay?

I believe Juju often uses bundles to deploy a set of charms AND relations together, so we'd want to make sure that the way we handle relations is compatible with that and that they won't need to resort to additional manual steps.

What's the out of order issue exactly?

Reply:
You shouldn't need to wait here before you do a `juju relate`. If you deploy from a bundle, the bundle is translated into a set of juju commands and the relations directly establish. 

If there is a charm side problem you're trying to mitigate with the `juju wait` it's most likely either a charm or a juju bug.

Reply:
Thanks @canonical/automation-and-tooling 

To sidestep the race, I changed the lxd side to process the client cert in relation-changed instead of relation-joined. This ensures the data put by the https-client unit is available when lxd is triggered.

Before that, I was seeing this:

|Unit|Message|
|---|---|
|unit-lxd-1|Emitting Juju event https_relation_created.| 
|unit-https-client-2|Emitting Juju event https_relation_created.| 
|unit-https-client-2|Client certificate uploaded to remote unit| 
|unit-lxd-1|Emitting Juju event https_relation_joined.| 
|unit-lxd-1|No version found in https-client/2| 
|unit-lxd-1|Emitting Juju event https_relation_changed.| 
|unit-lxd-1|Emitting Juju event https_relation_changed.| 
|unit-https-client-2|Emitting Juju event https_relation_joined.| 
|unit-https-client-2|Emitting Juju event https_relation_changed.| 
|unit-https-client-2|No version found in lxd/1| 

With the change, it now looks like:
|Unit|Message|
|---|---|
|unit-https-client-4|Emitting Juju event https_relation_created.|
|unit-https-client-4|Client certificate uploaded to remote unit|
|unit-https-client-4|Emitting Juju event https_relation_joined.|
|unit-lxd-2|Emitting Juju event https_relation_created.|
|unit-https-client-4|Emitting Juju event https_relation_changed.|
|unit-https-client-4|No version found in lxd/2|
|unit-lxd-2|Emitting Juju event https_relation_joined.|
|unit-lxd-2|Emitting Juju event https_relation_changed.|
|unit-lxd-2|The client certificate is now trusted| 
|unit-lxd-2|Connection information sent to https-client/4|
|unit-https-client-4|Emitting Juju event https_relation_changed.|
|unit-https-client-4|lxd/2 runs LXD version: 4.19|

So we see that the https-client unit received a notification too early but it was re-triggered which makes it race-free.